Update draft-ietf-cats-framework.md (#91)
boucadair/CATS-framework: draft-ietf-cats-framework.md

  : The same service can be provided in many locations; each of them constitutes a service instance.

Computing Service:

>   : An offering that is made available by a provider by orchestrating a set of computing resources to process application content.

Reviewer: Do we really need to mention "to process application content"?

Author: this is for sync up with the definition in use case draft. I am good with removing or keeping it. But if we want to address the comment, better to keep it. LOL. Only some editorial modification.

>   : Some differences and similarities exist between service and computing service:

Reviewer: "computing service" is still a service! Also, why are we adding this diff here? How this is linked to "application" mentioned right before?

Author: change it to service? or just delete it. As long as the definition is the same in use case and framework draft, I am ok.

>     * Service is packet-based, while computing service is transaction-based. Application content is exposed within a computing service.

Reviewer: Whether this is packet based or not actually depends on the service. Let alone that we need to define what is a transaction. We don't actually need to dive into these details from a CATS perspective.

Author: Personally, I agree with you. Do yo mind to simplify the definition? we need to sync up with the one in use case draft. Better to send an email in the CATS ML

 : A CATS-serviced site is a service site that is connected to a CATS-Forwarder.

Service contact instance:

>   : A client-facing service function instance that is responsible for receiving requests in the context of a given service. A service contact instance can represent one or more service instance. Steering beyond the service contact instance is hidden to both clients and CATS components. A service request is processed according to the service logic (e.g., handle locally or solicit backend resources).

Reviewer: ```suggestion
  : A client-facing service function instance that is responsible for receiving requests in the context of a given service.
  : A service contact instance can handle one or more service instances.
  : Steering beyond a service contact instance is hidden to both clients and CATS components.
  : A service request is processed according to the service logic (e.g., handle locally or solicit backend resources).
```

"A service contact instance can represent one or more service instance." seems to be redundant with "A service contact instance may dispatch service requests to one or more service instances". 

Also, I don't parse what is meant by "represent"? Do you mean "A service contact instance can handle one or more service instances."?

 : Service requests are not explicitly sent by clients to CATS-Forwarders.

CATS-Forwarder:

>  : A network entity that steers traffic specific to a service request towards a corresponding yet selected service contact instance according to the forwarding decisions made by the decision point based on CATS information. The decision point, which is on the C-PS, is responsible for making the steering policy, and it may or may not on the CATS-Forwarder.

Reviewer: ```suggestion
 : A network entity that steers traffic specific to a service request towards a corresponding yet selected service contact instance according to provisioned forwarding decisions. These decisions are supplied by a C-PS, which may or may not be on the CATS-Forwarder.
```

Reviewer: The tweaked text seems to conflict with this part "the choice of the service contact instance may be left open to the Egress CATS-Forwarder" in subsequent part of the text.

### Service Sites, Services Instances, and Service Contact Instances {#sec-service-sites}

Service sites are the premises that host a set of computing resources. As mentioned in {{cats-ids}}, a compute service (e.g., for face recognition purposes or a game server) is uniquely identified by a CATS Service IDentifier (CS-ID). The CS-ID does not need to be globally unique, though, for example, in the case of CATS working in a limited domain.

> Service instances are the accessed by service contact instance in CATS framework, therefore, a client will not see the service instances but only the service contact instance.

Reviewer: ```suggestion
As service instances are accessed via a service contact instance, a client will not see the service instances but only the service contact instance.
```

> A single service can be represented and accessed via several contact instances that run in different regions of a network.

Reviewer: ```suggestion
A single service can be represented and accessed via several contact instances that run in same or different regions of a network.
```

### Overlay CATS-Forwarders {#sec-ocr}

> The Egress CATS-Forwarders are the endpoints that behave as an overlay egress for service requests that are forwarded over a CATS infrastructure. A service site that hosts service instances may be connected to one or more Egress CATS-Forwarders (that is, multi-homing is of course a design option). If a C-PS has selected a specific service contact instance and the C-TC has marked the traffic with the CIS-ID related information, the Egress CATS-Forwarder then forwards traffic to the relevant service contact instance according the information. In some cases, the choice of the service  contact instance may be left open to the Egress CATS-Forwarder (i.e., traffic is marked only with the CS-ID). In such cases, the Egress CATS-Forwarder selects a service contact instance using its knowledge of service and network capabilities as well as the current load as observed by the CATS-Forwarder, among other considerations. Absent explicit policy, an Egress CATS-Forwarder must make sure to forward all packets that pertain to a given service request towards the same service contact instance.

Reviewer: ```suggestion
Egress CATS-Forwarders are the endpoints that behave as an overlay egress for service requests that are forwarded over a CATS infrastructure. A service site that hosts service instances may be connected to one or more Egress CATS-Forwarders (e.g., multi-homing design). If a C-PS has selected a specific service contact instance and the C-TC has marked the traffic with the CIS-ID related information, the Egress CATS-Forwarder then forwards traffic to the relevant service contact instance accordingly. In some cases, the choice of the service contact instance may be left open to the Egress CATS-Forwarder (i.e., traffic is marked only with the CS-ID). In such cases, the Egress CATS-Forwarder selects a service contact instance using its knowledge of service and network capabilities as well as the current load as observed by the CATS-Forwarder, among other considerations. Absent explicit policy, an Egress CATS-Forwarder must make sure to forward all packets that pertain to a given service request towards the same service contact instance.
```

Note that, depending on the design considerations and service requirements, per-service  contact instance computing-related metrics or aggregated per-site computing related metrics (and a combination thereof) can be used by a C-PS. Using aggregated per-site computing related metrics appears as a preferred option scalability-wise, but relies on Egress CATS-Forwarders that connect to various service contact instances to select the proper service contact instance. An Egress CATS-Forwarder may choose to aggregate the metrics from different sites as well. In this case, the Egress CATS-Forwarder will choose the best site by itself when the packets arrive at it.

### Underlay Infrastructure

The "underlay infrastructure" in {{fig-cats-components}} indicates an IP and/or MPLS network that is not necessarily CATS-aware. The CATS paths that are computed by a C-PS will be distributed among the CATS-Forwarders ({{sec-ocr}}), and will not affect the underlay nodes. Underlay nodes are typically P routers ({{Section 5.3.1 of ?RFC4026}}).

## Deployment Considerations

> This document does not make any assumption about how the various CATS functional elements are implemented and deployed. Concretely, whether a CATS deployment follows a fully distributed design or relies upon a mix of centralized (e.g., a C-PS) and distributed CATS functions (e.g., CATS traffic classifiers) is deployment-specific and may reflect the specific design of the (CATS) service provider.

Reviewer: ```suggestion
This document does not make any assumption about how the various CATS functional elements are implemented and deployed. Concretely, whether a CATS deployment follows a fully distributed design or relies upon a mix of centralized (e.g., a C-PS) and distributed CATS functions (e.g., CATS traffic classifiers) is deployment-specific and may reflect the savoir-faire of the (CATS) service provider.
```

Author: LOL, I try to find an English word to replace it, but it seems that I failed.  I am good with this.

A C-PS computes paths that lead to Egress CATS-Forwarders according to both service and network metrics that were advertised. A C-PS may be collocated with an Ingress CATS-Forwarder (as shown in {{fig-cats-example-overlay}}) or logically centralized (in a centralized model or hybrid model).

> This document does not specify any algorithm for path computation and selection purposes to be supported by C-PSes. However, it is expected that a service request or local policy may feed the C-PS computation logic with Objective Functions that provide some information about the path characteristics (e.g., in terms of maximum latency) and the preference in selecting the service contact instance.

Reviewer 2: ```suggestion
This document does not specify any path computation and instance selection algorithm running in C-PSes. However, it is expected that a service request or a policy may feed the C-PS computation logic with Objective Functions that provide the optimization goal or preference in selecting the service contact instance.
```

Reviewer: Any reason to remove the example ("e.g., in terms of maximum latency")? Otherwise, we have text that is too generic? I would keep the example here. Also, may delete "preference" as this can be part of the goal. 

I still prefer the OLD text.

Reviewer 2: > Any reason to remove the example ("e.g., in terms of maximum latency")? Otherwise, we have text that is too generic? I would keep the example here. Also, may delete "preference" as this can be part of the goal.
> 
> I still prefer the OLD text.

Objective Functions does not provide information of the path charateristics. The path characteristics is the input of the Objective Function.

Reviewer: I was not questioning that part of the modif :-)

Author: To me, we may need to use some straightforward wording for better understanding. Objective function introduces unnecessary complexities to me.
